Re-expose the reqwest method to add additional root certificates to the HTTP client

poljar opened the pull request:
This PR mainly allows clients to add root certificates to the HTTP client without having to build the `reqwest::Client` manually.

We also re-expose this functionality through the bindings.

- [x] Public API changes documented in changelogs (optional)

bnjbvr commented on bindings/matrix-sdk-ffi/src/authentication_service.rs:
```diff
@@ -176,10 +177,15 @@ impl HomeserverLoginDetails {
         base_path: String,
         passphrase: Option<String>,
         user_agent: Option<String>,
+        additional_root_certificates: Vec<Vec<u8>>,
```
In the FFI layer, could we use a type alias instead of `Vec<u8>` meaning a PEM or "der" certificate?

poljar replied:
Added: https://github.com/matrix-org/matrix-rust-sdk/pull/3126/commits/df1004fc181351b0ae553946df02468df53f9ca1

bnjbvr commented on bindings/matrix-sdk-ffi/src/client_builder.rs:
```diff
@@ -220,6 +230,25 @@ impl ClientBuilder {
+            if let Ok(cert) = Certificate::from_der(&certificate) {
+                certificates.push(cert);
+            } else {
+                match Certificate::from_pem(&certificate) {
+                    Ok(cert) => {
+                        certificates.push(cert);
+                    }
+                    Err(e) => anyhow::bail!("Failed to add a root certificate {e:?}"),
+                }
```
not sure if cuter or not, I'll let you decide :) 

```suggestion
                let cert = Certificate::from_pem(&certificate).unwrap_or_else(|e| anyhow::anyhow!("Failed to add a root certification {e}"))?;
                certificates.push(cert);
```

bnjbvr reviewed (changes requested):
Thank you! Is there a way to test this in the SDK?

poljar commented:
> Thank you! Is there a way to test this in the SDK?

Well we can't check if the cert was added to the `reqwest::Client`, that's internal state, what we could do is try to connect to a server with a self-signed certificate.

But... https://github.com/LukeMathWalker/wiremock-rs/issues/58

poljar commented:
I force pushed a rebase to get the conflict with main resolved, the commits stayed functionally the same and the changes addressing the review are still in a fixup commit.

bnjbvr reviewed (approved):
I see, thanks.

Is it worth to check, even superficially, if the inner reqwest client has some root certificates, if we added some in the SDK `Client`? It's a bit of a dumb test, and it would require that `reqwest` has some kind of getters for those root certificates, but it would help keeping this shiny high test-coverage percentage :smiling_face_with_tear:

poljar commented:
> I see, thanks.
> 
> Is it worth to check, even superficially, if the inner reqwest client has some root certificates, if we added some in the SDK `Client`? It's a bit of a dumb test, and it would require that `reqwest` has some kind of getters for those root certificates, but it would help keeping this shiny high test-coverage percentage 🥲

Yeah that's what I'm saying, it's internal state to the `reqwest::Client` I don't see how to check this: https://docs.rs/reqwest/latest/reqwest/struct.Client.html.
